File: src/org/cappuccino_project/ide/intellij/plugin/psi/types/ObjJElementTypeFactory.java
package org.cappuccino_project.ide.intellij.plugin.psi.types;

import com.intellij.psi.tree.IElementType;
import static org.cappuccino_project.ide.intellij.plugin.stubs.types.ObjJStubTypes.*;
import org.jetbrains.annotations.NotNull;

public interface ObjJElementTypeFactory {
    public static IElementType factory(@NotNull String name) {
        switch(name) {
            case "ObjJ_ACCESSOR_PROPERTY": return ACCESSOR_PROPERTY;
            case "ObjJ_FUNCTION_DECLARATION": return FUNCTION_DECLARATION;
            case "ObjJ_FUNCTION_LITERAL": return FUNCTION_LITERAL;
            case "ObjJ_GLOBAL_VARIABLE_DECLARATION": return GLOBAL_VARIABLE;
            case "ObjJ_IMPLEMENTATION_DECLARATION":return IMPLEMENTATION;
            case "ObjJ_IMPORT_FILE": return IMPORT_FILE;
            case "ObjJ_IMPORT_FRAMEWORK": return IMPORT_FRAMEWORK;
            case "ObjJ_INCLUDE_FILE": return INCLUDE_FILE;
            case "ObjJ_INCLUDE_FRAMEWORK": return INCLUDE_FRAMEWORK;
            case "ObjJ_INSTANCE_VARIABLE_DECLARATION": return INSTANCE_VAR;
            case "ObjJ_METHOD_CALL": return METHOD_CALL;
            case "ObjJ_METHOD_HEADER":return METHOD_HEADER;
            case "ObjJ_PREPROCESSOR_DEFINE_FUNCTION": return PREPROCESSOR_FUNCTION;
            case "ObjJ_PROTOCOL_DECLARATION":return PROTOCOL;
            case "ObjJ_SELECTOR_LITERAL": return SELECTOR_LITERAL;
            case "ObjJ_VAR_TYPE_ID": return VAR_TYPE_ID;
            default:
                throw new RuntimeException("Failed to find element type in factory for type <"+name+">");
        }
    }
}

File: src/org/cappuccino_project/ide/intellij/plugin/psi/types/ObjJElementTypeFactory.kt
package org.cappuccino_project.ide.intellij.plugin.psi.types

import com.intellij.psi.tree.IElementType
import org.cappuccino_project.ide.intellij.plugin.stubs.types.ObjJStubTypes.*

interface ObjJElementTypeFactory {
    companion object {
        fun factory(name: String): IElementType {
            when (name) {
                "ObjJ_ACCESSOR_PROPERTY" -> return ACCESSOR_PROPERTY
                "ObjJ_FUNCTION_DECLARATION" -> return FUNCTION_DECLARATION
                "ObjJ_FUNCTION_LITERAL" -> return FUNCTION_LITERAL
                "ObjJ_GLOBAL_VARIABLE_DECLARATION" -> return GLOBAL_VARIABLE
                "ObjJ_IMPLEMENTATION_DECLARATION" -> return IMPLEMENTATION
                "ObjJ_IMPORT_FILE" -> return IMPORT_FILE
                "ObjJ_IMPORT_FRAMEWORK" -> return IMPORT_FRAMEWORK
                "ObjJ_INCLUDE_FILE" -> return INCLUDE_FILE
                "ObjJ_INCLUDE_FRAMEWORK" -> return INCLUDE_FRAMEWORK
                "ObjJ_INSTANCE_VARIABLE_DECLARATION" -> return INSTANCE_VAR
                "ObjJ_METHOD_CALL" -> return METHOD_CALL
                "ObjJ_METHOD_HEADER" -> return METHOD_HEADER
                "ObjJ_PREPROCESSOR_DEFINE_FUNCTION" -> return PREPROCESSOR_FUNCTION
                "ObjJ_PROTOCOL_DECLARATION" -> return PROTOCOL
                "ObjJ_SELECTOR_LITERAL" -> return SELECTOR_LITERAL
                "ObjJ_VAR_TYPE_ID" -> return VAR_TYPE_ID
                "ObjJ_VARIABLE_NAME" -> return VARIABLE_NAME
                else -> throw RuntimeException("Failed to find element type in factory for type <$name>")
            }
        }
    }
}
